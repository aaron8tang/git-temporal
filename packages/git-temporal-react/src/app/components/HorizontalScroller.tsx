import React from 'react';
import { style } from 'app/styles';
import { relative } from 'path';
import { debounce } from 'app/utilities/debounce';

import { CaretRightIcon } from 'app/components/CaretRightIcon';
import { CaretLeftIcon } from 'app/components/CaretLeftIcon';

export interface HorizontalScrollerProps {
  // The children are the menu content
  children: string | JSX.Element | JSX.Element[];
  scrollLeft?: number;
  style?: string | object;
  innerStyle?: string | object;
  iconStyle?: string | object;
  onScroll?: (scrollLeft: number) => void;
}

const initialState = {
  showingLeftTouch: true,
  showingRightTouch: true,
};

type HorizontalScrollerState = Readonly<typeof initialState>;

const farOuterStyle = {
  display: 'flex',
  width: '100%',
  position: 'relative',
};

const outerStyle = {
  _extends: 'fill',
  overflowX: 'scroll',
  position: relative,
  scrollBehavior: 'smooth',
};

const innerContainerStyle = {
  position: 'relative',
  display: 'block',
};

const touchArea = {
  position: 'absolute',
  top: 0,
  width: 40,
  height: '100%',
  verticalAlign: 'middle',
  zIndex: 2,
  marginTop: 5,
  opacity: 0.5,
  color: 'lightskyblue',
};

const leftTouchStyle = {
  _extends: touchArea,
  left: -20,
  // borderTop: '60px solid transparent',
  // borderBottom: '60px solid transparent',
  // borderRight: '20px solid rgba(96, 96, 96, .6)',
};

const rightTouchStyle = {
  _extends: touchArea,
  right: -20,
  // borderTop: '60px solid transparent',
  // borderBottom: '60px solid transparent',
  // borderLeft: '20px solid rgba(96, 96, 96, .6)',
};

const iconStyle = {
  transform: 'scaleY(3)',
  maxHeight: '50%',
  top: '12%',
  position: 'absolute',
};

export class HorizontalScroller extends React.Component<
  HorizontalScrollerProps,
  HorizontalScrollerState
> {
  readonly state: HorizontalScrollerState = initialState;
  private innerContainerRef;
  private outerContainerRef;
  private debouncedOnScroll;

  constructor(props) {
    super(props);
    this.innerContainerRef = React.createRef();
    this.outerContainerRef = React.createRef();

    this.onScroll = this.onScroll.bind(this);
    this.debouncedOnScroll = debounce(this.onScroll, 50);
  }

  componentDidMount() {
    this.showHideTouchControls();
  }

  componentDidUpdate(prevProps) {
    this.showHideTouchControls();
    if (prevProps.scrollLeft !== this.props.scrollLeft) {
      this.outerContainerRef.current.scrollLeft = this.props.scrollLeft;
    }
  }

  render() {
    return (
      <div style={style(farOuterStyle)}>
        <div
          style={style(outerStyle, this.props.style)}
          ref={this.outerContainerRef}
          onScroll={this.debouncedOnScroll}
          data-testid="outerContainer"
        >
          <div
            style={style(innerContainerStyle, this.props.innerStyle)}
            ref={this.innerContainerRef}
            data-testid="innerContainer"
          >
            {this.props.children}
          </div>
        </div>
        {this.state.showingLeftTouch && (
          <div
            style={style(leftTouchStyle)}
            onClick={this.onLeftTouchClick}
            onDoubleClick={this.onLeftTouchDoubleClick}
            data-testid="leftTouch"
          >
            <CaretLeftIcon
              width={40}
              height={100}
              style={style(iconStyle, this.props.iconStyle)}
            />{' '}
          </div>
        )}
        {this.state.showingRightTouch && (
          <div
            style={style(rightTouchStyle)}
            onClick={this.onRightTouchClick}
            onDoubleClick={this.onRightTouchDoubleClick}
            data-testid="rightTouch"
          >
            <CaretRightIcon
              width={40}
              height={100}
              style={style(iconStyle, this.props.iconStyle)}
            />
          </div>
        )}
      </div>
    );
  }

  private showHideTouchControls() {
    const innerEl = this.innerContainerRef.current;
    const outerEl = this.outerContainerRef.current;

    const innerBoundingRect = innerEl.getBoundingClientRect();
    const showingLeftTouch = outerEl.scrollLeft > 0;
    const showingRightTouch =
      innerEl.scrollWidth > outerEl.clientWidth &&
      outerEl.scrollLeft + innerBoundingRect.width < innerEl.scrollWidth;
    const shouldUpdate =
      showingLeftTouch !== this.state.showingLeftTouch ||
      showingRightTouch !== this.state.showingRightTouch;
    if (shouldUpdate) {
      this.setState({ showingLeftTouch, showingRightTouch });
    }
    // console.log('innerBoundingRect', innerBoundingRect);
    // console.log('innerEl.scrollWidth', innerEl.scrollWidth);
    // console.log('outerEl.clientWidth', outerEl.clientWidth);
    // console.log('outerEl.scrollLeft', outerEl.scrollLeft);
    // console.log('showingLeftTouch', showingLeftTouch);
    // console.log('showingRightTouch', showingRightTouch);
  }

  private onScroll() {
    this.showHideTouchControls();
    this.props.onScroll &&
      this.props.onScroll(this.outerContainerRef.current.scrollLeft);
  }

  private onLeftTouchClick = () => {
    const innerEl = this.innerContainerRef.current;
    const outerEl = this.outerContainerRef.current;
    const innerBoundingRect = innerEl.getBoundingClientRect();

    let newScrollLeft = outerEl.scrollLeft - innerBoundingRect.width + 50;
    if (newScrollLeft < 0) {
      newScrollLeft = 0;
    }
    outerEl.scrollLeft = newScrollLeft;
  };

  private onLeftTouchDoubleClick = () => {
    const outerEl = this.outerContainerRef.current;
    outerEl.scrollLeft = 0;
  };

  private onRightTouchClick = () => {
    const innerEl = this.innerContainerRef.current;
    const outerEl = this.outerContainerRef.current;
    const innerBoundingRect = innerEl.getBoundingClientRect();
    const maxScrollLeft = innerEl.scrollWidth - innerBoundingRect.width;

    let newScrollLeft = outerEl.scrollLeft + innerBoundingRect.width - 50;
    if (newScrollLeft > maxScrollLeft) {
      newScrollLeft = maxScrollLeft;
    }
    outerEl.scrollLeft = newScrollLeft;
  };

  private onRightTouchDoubleClick = () => {
    const innerEl = this.innerContainerRef.current;
    const outerEl = this.outerContainerRef.current;
    const innerBoundingRect = innerEl.getBoundingClientRect();
    const maxScrollLeft = innerEl.scrollWidth - innerBoundingRect.width;

    outerEl.scrollLeft = maxScrollLeft;
  };
}
