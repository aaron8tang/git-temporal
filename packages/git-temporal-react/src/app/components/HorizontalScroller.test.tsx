import React from 'react';
import { mount } from 'enzyme';

import { HorizontalScroller } from './HorizontalScroller';

describe('components/HorizontalScroller', () => {
  describe('when rendered with 100% width div', () => {
    let wrapper;
    const onScrollMock = jest.fn();
    beforeAll(() => {
      wrapper = mount(
        <HorizontalScroller onScroll={onScrollMock}>
          <div style={{ width: '100%' }}>This is a 100% width div</div>
        </HorizontalScroller>
      );
    });

    test('it should match snapshot (it should have render just the children without left or right touch controls)")', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('when rendered with 200% width div', () => {
    let wrapper;
    const original__getBoundClientRect =
      Element.prototype.getBoundingClientRect;
    const onScrollMock = jest.fn();
    beforeAll(() => {
      Element.prototype.getBoundingClientRect;
      Element.prototype.getBoundingClientRect = jest.fn(() => {
        return {
          width: 400,
          height: 120,
          top: 0,
          left: 0,
          bottom: 0,
          right: 0,
        };
      });
      wrapper = mount(
        <HorizontalScroller onScroll={onScrollMock}>
          <div style={{ width: '200%' }}>This is a 200% width div</div>
        </HorizontalScroller>
      );
      const innerEl = wrapper.find('[data-testid="innerContainer"]');
      const outerEl = wrapper.find('[data-testid="outerContainer"]');
      // TODO : the below never changes the element properties in the test
      //   uncomment the console.log lines in showHideTouchControls() to see
      innerEl.scrollWidth = 200;
      outerEl.clientWidth = 400;
    });
    afterAll(() => {
      Element.prototype.getBoundingClientRect = original__getBoundClientRect;
    });

    test('it should match snapshot (should have rendered the child with a right touch control))', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('and then changing the scrollLeft, it should NOT have called the onScroll callback', () => {
      wrapper.setProps({ scrollLeft: 100 });
      // we should only call the onScroll callback for user initiate scroll events\\
      expect(onScrollMock).toHaveBeenCalledTimes(0);
    });

    //  TODO : this is a blatent hack at improving coverage, fix when we figureout how to
    //   set scrollLeft, scrollWidth, clientWidth on elements rendered with enzyme
    test('and then clicking the right touch, it should should scroll left', () => {
      wrapper.instance().onRightTouchClick();
      wrapper.instance().onRightTouchDoubleClick();
      wrapper.instance().onLeftTouchClick();
      wrapper.instance().onLeftTouchDoubleClick();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
