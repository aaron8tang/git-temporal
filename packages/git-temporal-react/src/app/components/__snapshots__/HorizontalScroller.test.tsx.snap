// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/HorizontalScroller when rendered with 100% width div it should match snapshot (it should have render just the children without left or right touch controls)") 1`] = `
<HorizontalScroller
  onScroll={[MockFunction]}
>
  <div
    data-testid="outerContainer"
    onScroll={[Function]}
    style={
      Object {
        "height": "100%",
        "overflowX": "scroll",
        "position": [Function],
        "width": "100%",
      }
    }
  >
    <div
      data-testid="innerContainer"
      style={
        Object {
          "display": "block",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "width": "100%",
          }
        }
      >
        This is a 100% width div
      </div>
    </div>
  </div>
</HorizontalScroller>
`;

exports[`components/HorizontalScroller when rendered with 200% width div and then clicking the right touch, it should should scroll left 1`] = `
<HorizontalScroller
  onScroll={[MockFunction]}
  scrollLeft={100}
>
  <div
    data-testid="outerContainer"
    onScroll={[Function]}
    style={
      Object {
        "height": "100%",
        "overflowX": "scroll",
        "position": [Function],
        "width": "100%",
      }
    }
  >
    <div
      data-testid="innerContainer"
      style={
        Object {
          "display": "block",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "width": "200%",
          }
        }
      >
        This is a 200% width div
      </div>
    </div>
  </div>
</HorizontalScroller>
`;

exports[`components/HorizontalScroller when rendered with 200% width div it should match snapshot (should have rendered the child with a right touch control)) 1`] = `
<HorizontalScroller
  onScroll={[MockFunction]}
>
  <div
    data-testid="outerContainer"
    onScroll={[Function]}
    style={
      Object {
        "height": "100%",
        "overflowX": "scroll",
        "position": [Function],
        "width": "100%",
      }
    }
  >
    <div
      data-testid="innerContainer"
      style={
        Object {
          "display": "block",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "width": "200%",
          }
        }
      >
        This is a 200% width div
      </div>
    </div>
  </div>
</HorizontalScroller>
`;
